### ISSUE TITLE
Issue with type UpdateData<T> from firebase/firestore 

### ISSUE THREAD

**Issue by nwaughachukwuma** [firebase/firebase-js-sdk#6105]
### [REQUIRED] Describe your environment

  * Operating System version: Mac OS
  * Browser version: N/A
  * Firebase SDK version: "firebase": "9.6.8", "firebase-admin": "10.0.2", "@firebase/rules-unit-testing": "2.0.2",
  * Firebase Product: Firestore

#### Steps to reproduce:
There is an issue with type `UpdateData<T>` imported from `firebase/firestore`, which fails to properly parse object type when data is nested deeper than one level - i.e. it passes `foo.bar` but not `foo.bar.baz`. Given the data shape and ops in the `relevant code` section, I get the following error on my IDE and also when I build my project: 

```bash
Argument of type '{ [x: string]: UserRole; }' is not assignable to parameter of type '{ members?: FieldValue | { [x: string]: FieldValue | { role?: UserRole | FieldValue | undefined; status?: MemberStatus | FieldValue | undefined; } | undefined; } | undefined; name?: string | ... 1 more ... | undefined; } & AddPrefixToKeys<...>'.
  Type '{ [x: string]: UserRole; }' is not assignable to type 'AddPrefixToKeys<"members", { [x: string]: FieldValue | { role?: UserRole | FieldValue | undefined; status?: MemberStatus | FieldValue | undefined; } | undefined; }>'.
    'string' and '`members.${string}`' index signatures are incompatible.
      Type 'UserRole' is not assignable to type 'FieldValue | { role?: UserRole | FieldValue | undefined; status?: MemberStatus | FieldValue | undefined; } | undefined'.
```

#### Relevant Code:

```ts
type UserID = string
enum UserRole {
  Admin = 'Admin',
  Member = 'Member',
}
type MemberStatus = 'pending' | 'active'
interface TeamMember {
  role: UserRole
  status: MemberStatus
}

/** 
* this works
*/
interface Team {
  name: string
  members: Record<UserID, UserRole>
}
Update Op:
~.update({ [`members.${uid}`]: UserRole.Member })

/** 
* this doesn't work
*/
interface Team {
  name: string
  members: Record<UserID, TeamMember> 
}
Update Op:
~.update({ [`members.${uid}.role`]: UserRole.Member })
```

**Comment by harshmandan** [comment]
I have the same issue.

**Comment by ebeloded** [comment]
Any updates on this issue? The current implementation of `UpdateData` type doesn't allow working with complex objects in a type-safe manner.

**Comment by ehsannas** [comment]
Hi everyone, sorry for the delay. I'll try to prioritize the investigation for this issue.

**Comment by ehsannas** [comment]
Hi everyone, sorry for the wait and for the trouble. Do you currently have a workaround for this (e.g. using `any`)?

**Comment by ebeloded** [comment]
Hi @ehsannas. Thank you for check-in. Currently, we cover up some parts of data interfaces with `any` with an intention to update when the issue is resolved.

**Comment by nwaughachukwuma** [comment]
Same.

**Comment by nwaughachukwuma** [comment]
Hi @ehsannas, I opened a PR in #6371 to fix this issue. Please follow-up

**Comment by dconeybe** [comment]
This is a duplicate of https://github.com/firebase/firebase-js-sdk/issues/5853. Please follow that issue for updates.

**Comment by dconeybe** [comment]
This is fixed in the v10.0.0 release: https://firebase.google.com/support/release-notes/js#version_1000_-_july_6_2023
